internal/character/clara/ult.go

```diff
@@ -0,0 +1,79 @@
+package clara
+
+import (
+	"github.com/simimpact/srsim/pkg/engine/event"
+	"github.com/simimpact/srsim/pkg/engine/info"
+	"github.com/simimpact/srsim/pkg/engine/modifier"
+	"github.com/simimpact/srsim/pkg/engine/prop"
+	"github.com/simimpact/srsim/pkg/key"
+	"github.com/simimpact/srsim/pkg/model"
+)
+
+const (
+	Ult        key.Modifier = "clara-ult" // MAvatar_Klara_00_Ultra_WarriorMode
+	UltCounter key.Modifier = "clara-ult-enhanced-counter"
+)
+
+type ultCounterNum int
+
+func init() {
+	modifier.Register(Ult, modifier.Config{
+		BehaviorFlags: []model.BehaviorFlag{model.BehaviorFlag_STAT_BURST},
+		Stacking:      modifier.Refresh,
+		Duration:      2,
+		Listeners: modifier.Listeners{
+			OnAdd: func(mod *modifier.Instance) {
+				state := mod.State().(State)
+
+				// extra res
+				mod.AddProperty(prop.AllDamageReduce, ultCut[state.ultLevelIndex])
+				// extra aggro
+				mod.AddProperty(prop.AggroPercent, 5)
+
+				amt := 2
+				if mod.Engine().HasModifier(mod.Owner(), E6) {
+					amt = 3
+				}
+
+				mod.Engine().AddModifier(mod.Owner(), info.Modifier{
+					Name:   UltCounter,
+					Source: mod.Owner(),
+					State:  ultCounterNum(amt),
+				})
+			},
+		},
+	})
+
+	// mod check in talent.go
+	modifier.Register(UltCounter, modifier.Config{
+		Listeners: modifier.Listeners{
+			OnAfterHit: func(mod *modifier.Instance, e event.HitEnd) {
+				// after counter
+				if e.AttackType == model.AttackType_INSERT {
+					remaining := mod.State().(ultCounterNum)
+					remaining -= 1
+					if remaining <= 0 { // executes 2 counters
+						mod.RemoveSelf()
+					}
+				}
+			},
+		},
+	})
```

Comment:
This is very wrong. This will have UltCounter reach 0 after first ult insert (since 2-hit attack, will have remaining reach 0).

This modifier should have no listener, just use the modifier `Count` system instead. This config then should only specify a max count of 3 (since `MaxCount` is required when using count). When adding the `UltCounter` modifier, set the starting count to 2 or 3 depending on E6 status

Modifier also needs stacking behavior `ReplaceBySource` so count gets reset if it is non-zero when you ult